File: lib/utils/forecast.ts
// Compound net-worth forecast — "at this pace, when do I get there, and
// what would it take to get there sooner?"
//
// Everything runs through ONE monthly simulation rather than closed-form
// annuity algebra: it stays honest when the contribution grows each year,
// when the return is zero or negative, and it makes the inverse questions
// (required saving, required return) a bisection over the same loop, so the
// forward and inverse answers can never disagree with each other.
//
// Model, per month:   nw ← nw × (1 + r_m) + saving
//                     saving grows by `contributionGrowthPct` once a year
// where r_m = (1 + annualReturnPct/100)^(1/12) − 1 (geometric, so a 7% year
// really compounds to 7%). Contributions land at month end.
//
// The iOS app carries a line-for-line port (Native/Forecast.swift); keep the
// two in step — they are meant to be diffed by eye.

export interface ForecastInputs {
  /** Net worth today, display currency. */
  netWorth: number;
  /** Money added per month (income − expenses), display currency. */
  monthlySaving: number;
  /** Long-run annual return on the whole net worth, percent (7 = 7%). */
  annualReturnPct: number;
  /** How much the monthly saving grows each year, percent (raises, etc). */
  contributionGrowthPct: number;
}

/** Hard horizon: past this the answer is "not on this path", not a date. */
export const FORECAST_MAX_MONTHS = 100 * 12;

export function monthlyRate(annualReturnPct: number): number {
  return Math.pow(1 + annualReturnPct / 100, 1 / 12) - 1;
}

/**
 * Net worth after each of `months` months (index 0 = today), plus the
 * "savings only" path so a chart can show what compounding contributes.
 */
export function projectPath(
  inputs: ForecastInputs,
  months: number,
): { withGrowth: number[]; savingsOnly: number[] } {
  const r = monthlyRate(inputs.annualReturnPct);
  const withGrowth: number[] = [inputs.netWorth];
  const savingsOnly: number[] = [inputs.netWorth];
  let nw = inputs.netWorth;
  let flat = inputs.netWorth;
  let saving = inputs.monthlySaving;
  for (let m = 1; m <= months; m++) {
    if (m > 1 && (m - 1) % 12 === 0) saving *= 1 + inputs.contributionGrowthPct / 100;
    nw = nw * (1 + r) + saving;
    flat = flat + saving;
    withGrowth.push(nw);
    savingsOnly.push(flat);
  }
  return { withGrowth, savingsOnly };
}

/**
 * Months until net worth first reaches `target`; 0 when already there; null
 * when it never happens inside the horizon (shrinking or flat path).
 */
export function monthsToReach(inputs: ForecastInputs, target: number): number | null {
  if (inputs.netWorth >= target) return 0;
  const r = monthlyRate(inputs.annualReturnPct);
  let nw = inputs.netWorth;
  let saving = inputs.monthlySaving;
  for (let m = 1; m <= FORECAST_MAX_MONTHS; m++) {
    if (m > 1 && (m - 1) % 12 === 0) saving *= 1 + inputs.contributionGrowthPct / 100;
    nw = nw * (1 + r) + saving;
    if (nw >= target) return m;
  }
  return null;
}

/**
 * Smallest monthly saving that reaches `target` within `months` at the given
 * return — the "what would I need to put away" answer. Null when even an
 * absurd saving can't (months ≤ 0).
 */
export function requiredMonthlySaving(
  base: Omit<ForecastInputs, "monthlySaving">,
  target: number,
  months: number,
): number | null {
  if (months <= 0) return null;
  const reaches = (saving: number) =>
    (monthsToReach({ ...base, monthlySaving: saving }, target) ?? Infinity) <= months;
  if (reaches(0)) return 0;
  let lo = 0;
  let hi = Math.max(1, target); // saving the whole target monthly always works
  if (!reaches(hi)) return null;
  for (let i = 0; i < 60; i++) {
    const mid = (lo + hi) / 2;
    if (reaches(mid)) hi = mid; else lo = mid;
  }
  return hi;
}

/**
 * Smallest annual return (percent) that reaches `target` within `months` at
 * the current saving. Null when nothing inside −50%…+100% does.
 */
export function requiredAnnualReturn(
  base: Omit<ForecastInputs, "annualReturnPct">,
  target: number,
  months: number,
): number | null {
  if (months <= 0) return null;
  const reaches = (pct: number) =>
    (monthsToReach({ ...base, annualReturnPct: pct }, target) ?? Infinity) <= months;
  let lo = -50;
  let hi = 100;
  if (reaches(lo)) return lo;
  if (!reaches(hi)) return null;
  for (let i = 0; i < 60; i++) {
    const mid = (lo + hi) / 2;
    if (reaches(mid)) hi = mid; else lo = mid;
  }
  return hi;
}

/** "14y 7m", "8 months", "3 weeks" — a duration people say out loud. */
export function describeMonths(months: number): string {
  if (months <= 0) return "now";
  if (months < 1) return "under a month";
  const y = Math.floor(months / 12);
  const m = Math.round(months % 12);
  if (y === 0) return `${m} month${m === 1 ? "" : "s"}`;
  if (m === 0) return `${y} year${y === 1 ? "" : "s"}`;
  return `${y}y ${m}m`;
}

/**
 * The measured "pace" this app is built to report honestly: how fast net
 * worth grew BEYOND what was deposited — i.e. what the assets themselves
 * did. Annualised. Null when the window is under 90 days (a quarter of
 * one market swing is not a pace) or the numbers are unusable.
 *
 *   growth = nwEnd − nwStart − netSavings(window)
 *   pace   = growth / avg(nwStart, nwEnd) × 365 / days
 */
export function measuredAnnualPacePct(args: {
  nwStart: number;
  nwEnd: number;
  netSavingsInWindow: number;
  windowDays: number;
}): number | null {
  const { nwStart, nwEnd, netSavingsInWindow, windowDays } = args;
  if (windowDays < 90) return null;
  const avg = (nwStart + nwEnd) / 2;
  if (!(avg > 0)) return null;
  const growth = nwEnd - nwStart - netSavingsInWindow;
  const pct = (growth / avg) * (365 / windowDays) * 100;
  return Number.isFinite(pct) ? pct : null;
}

/**
 * Average monthly saving from the ledgers. Only months where BOTH income
 * and expenses were recorded count — a month with income and zero expenses
 * is a month before expense tracking started, not a month of perfect
 * frugality, and it would inflate the pace. Falls back to every month with
 * income when fewer than two such months exist.
 */
export function measuredMonthlySaving(
  monthly: { income: number; expenses: number }[],
): number | null {
  const complete = monthly.filter((m) => m.income > 0 && m.expenses > 0);
  const pool = complete.length >= 2 ? complete : monthly.filter((m) => m.income > 0);
  if (pool.length === 0) return null;
  return pool.reduce((s, m) => s + (m.income - m.expenses), 0) / pool.length;
}

/** Shared, synced assumptions (app_data key `forecast_assumptions`). */
export interface ForecastAssumptions {
  /** null = use the measured pace when available, else the balanced default. */
  annualReturnPct: number | null;
  /** null = use the measured monthly saving. */
  monthlySaving: number | null;
  contributionGrowthPct: number;
}

export const DEFAULT_ASSUMPTIONS: ForecastAssumptions = {
  annualReturnPct: null,
  monthlySaving: null,
  contributionGrowthPct: 0,
};

/** The long-run assumptions offered beside the measured pace. */
export const RETURN_PRESETS: { label: string; pct: number; note: string }[] = [
  { label: "Cautious", pct: 4, note: "bonds-heavy, or a rough decade" },
  { label: "Balanced", pct: 7, note: "long-run diversified equities" },
  { label: "Aggressive", pct: 10, note: "all-in growth, in a good era" },
];

/** Balanced is the fallback when nothing is measured yet. */
export const FALLBACK_RETURN_PCT = 7;

/**
 * The measured pace becomes the DEFAULT return only with a full year of
 * history behind it. Under that it is offered as a chip, never assumed —
 * a short window annualised is a mood, not a law.
 */
export const MEASURED_PACE_MIN_DAYS = 365;

File: android/app/src/main/java/com/piyawatpm/vesta/core/Forecast.kt
package com.piyawatpm.vesta.core

import com.piyawatpm.vesta.data.ForecastAssumptions
import com.piyawatpm.vesta.data.NetworthGoal
import com.piyawatpm.vesta.data.VestaStore
import java.time.LocalDate
import java.time.YearMonth
import kotlin.math.abs
import kotlin.math.pow

// Compound net-worth forecast — line-for-line port of lib/utils/forecast.ts
// (which carries the test suite). One monthly simulation answers every
// question: forward ("when do I get there?") and inverse ("what saving /
// return would get me there by then?", by bisection over the same walk), so
// the answers can never disagree with each other.

data class ForecastInputs(
    val netWorth: Double,
    val monthlySaving: Double,
    val annualReturnPct: Double,
    val contributionGrowthPct: Double,
)

object ForecastMath {
    const val MAX_MONTHS = 100 * 12

    fun monthlyRate(annualPct: Double): Double = (1 + annualPct / 100).pow(1.0 / 12) - 1

    /** Net worth after each month (index 0 = today) with and without growth. */
    fun projectPath(inputs: ForecastInputs, months: Int): Pair<List<Double>, List<Double>> {
        val r = monthlyRate(inputs.annualReturnPct)
        val withGrowth = ArrayList<Double>(months + 1).apply { add(inputs.netWorth) }
        val savingsOnly = ArrayList<Double>(months + 1).apply { add(inputs.netWorth) }
        var nw = inputs.netWorth
        var flat = inputs.netWorth
        var saving = inputs.monthlySaving
        for (m in 1..months) {
            if (m > 1 && (m - 1) % 12 == 0) saving *= 1 + inputs.contributionGrowthPct / 100
            nw = nw * (1 + r) + saving
            flat += saving
            withGrowth.add(nw)
            savingsOnly.add(flat)
        }
        return withGrowth to savingsOnly
    }

    /** Months until `target` is first reached; 0 if already there; null if
     *  the path never gets there inside the horizon. */
    fun monthsToReach(inputs: ForecastInputs, target: Double): Int? {
        if (inputs.netWorth >= target) return 0
        val r = monthlyRate(inputs.annualReturnPct)
        var nw = inputs.netWorth
        var saving = inputs.monthlySaving
        for (m in 1..MAX_MONTHS) {
            if (m > 1 && (m - 1) % 12 == 0) saving *= 1 + inputs.contributionGrowthPct / 100
            nw = nw * (1 + r) + saving
            if (nw >= target) return m
        }
        return null
    }

    fun requiredMonthlySaving(
        netWorth: Double,
        annualReturnPct: Double,
        contributionGrowthPct: Double,
        target: Double,
        months: Int,
    ): Double? {
        if (months <= 0) return null
        fun reaches(saving: Double): Boolean {
            val inputs = ForecastInputs(netWorth, saving, annualReturnPct, contributionGrowthPct)
            return (monthsToReach(inputs, target) ?: Int.MAX_VALUE) <= months
        }
        if (reaches(0.0)) return 0.0
        var lo = 0.0
        var hi = maxOf(1.0, target)
        if (!reaches(hi)) return null
        repeat(60) {
            val mid = (lo + hi) / 2
            if (reaches(mid)) hi = mid else lo = mid
        }
        return hi
    }

    fun requiredAnnualReturn(
        netWorth: Double,
        monthlySaving: Double,
        contributionGrowthPct: Double,
        target: Double,
        months: Int,
    ): Double? {
        if (months <= 0) return null
        fun reaches(pct: Double): Boolean {
            val inputs = ForecastInputs(netWorth, monthlySaving, pct, contributionGrowthPct)
            return (monthsToReach(inputs, target) ?: Int.MAX_VALUE) <= months
        }
        var lo = -50.0
        var hi = 100.0
        if (reaches(lo)) return lo
        if (!reaches(hi)) return null
        repeat(60) {
            val mid = (lo + hi) / 2
            if (reaches(mid)) hi = mid else lo = mid
        }
        return hi
    }

    fun describe(months: Int): String {
        if (months <= 0) return "now"
        val y = months / 12
        val m = months % 12
        if (y == 0) return "$m month${if (m == 1) "" else "s"}"
        if (m == 0) return "$y year${if (y == 1) "" else "s"}"
        return "${y}y ${m}m"
    }

    /** Growth of net worth beyond deposits, annualized; null under 90 days. */
    fun measuredAnnualPacePct(
        nwStart: Double,
        nwEnd: Double,
        netSavings: Double,
        windowDays: Double,
    ): Double? {
        if (windowDays < 90) return null
        val avg = (nwStart + nwEnd) / 2
        if (avg <= 0) return null
        val pct = (nwEnd - nwStart - netSavings) / avg * (365 / windowDays) * 100
        return if (pct.isFinite()) pct else null
    }

    /** Only months with BOTH income and expenses logged count — a month with
     *  income and zero expenses predates expense tracking and would inflate
     *  the pace. Falls back to income months when fewer than two qualify. */
    fun measuredMonthlySaving(monthly: List<Pair<Double, Double>>): Double? {
        val complete = monthly.filter { it.first > 0 && it.second > 0 }
        val pool = if (complete.size >= 2) complete else monthly.filter { it.first > 0 }
        if (pool.isEmpty()) return null
        return pool.sumOf { it.first - it.second } / pool.size
    }

    const val FALLBACK_RETURN_PCT = 7.0

    data class Preset(val label: String, val pct: Double, val note: String)

    val presets = listOf(
        Preset("Cautious", 4.0, "bonds-heavy, or a rough decade"),
        Preset("Balanced", 7.0, "long-run diversified equities"),
        Preset("Aggressive", 10.0, "all-in growth, in a good era"),
    )

    fun addMonths(months: Int): LocalDate = LocalDate.now(SydneyTime.zone).plusMonths(months.toLong())

    fun monthYear(date: LocalDate): String =
        "${SydneyTime.monthShortName(date.monthValue)} ${date.year}"

    fun monthsUntil(ymd: String): Int? {
        val target = try { LocalDate.parse(ymd.take(10)) } catch (_: Exception) { return null }
        val today = LocalDate.now(SydneyTime.zone)
        return (YearMonth.from(target).let { t ->
            val base = YearMonth.from(today)
            (t.year - base.year) * 12 + (t.monthValue - base.monthValue)
        })
    }
}

// MARK: - Measured inputs from the store (ios DataStore extension port)

data class MeasuredPace(
    val monthlySaving: Double?,
    val pacePct: Double?,
    val paceDays: Double,
)

/** The measured pace is a trustworthy DEFAULT only once a full year is
 *  behind it — four months of a crypto dip annualised would otherwise
 *  headline a 15-year forecast. */
const val MEASURED_PACE_MIN_DAYS = 365.0

/** The honest inputs: saving from the ledgers, pace from the net-worth
 *  history — both in the display currency. Same definitions as the web. */
val VestaStore.measuredPace: MeasuredPace
    get() {
        // Last six complete months (exclude the current, partial one).
        val months = FlowMath.monthKeys(7).dropLast(1)
        val monthly = months.map { key ->
            monthTotal(income, key) to monthTotalExpenses(key)
        }
        val saving = ForecastMath.measuredMonthlySaving(monthly)

        // Pace: latest reading vs ~180 days back (or as far as history goes).
        var pace: Double? = null
        var days = 0.0
        val series = networthParsed
        val last = series.lastOrNull()
        if (last != null && series.size > 1) {
            val wanted = last.date - 180L * 86400_000L
            val start = series.firstOrNull { it.date >= wanted } ?: series[0]
            days = (last.date - start.date) / 86400_000.0
            val startKey = sydneyDay(start.date.toDouble())
            val endKey = sydneyDay(last.date.toDouble())
            val earned = income
                .filter { it.date >= startKey && it.date <= endKey }
                .sumOf { convert(it.amount, it.currency) }
            val spent = expenses
                .filter { it.date >= startKey && it.date <= endKey }
                .sumOf { convert(it.amount, it.currency) }
            pace = ForecastMath.measuredAnnualPacePct(
                nwStart = convert(start.valueUsd, "USD"),
                nwEnd = convert(last.valueUsd, "USD"),
                netSavings = earned - spent,
                windowDays = days,
            )
        }
        return MeasuredPace(saving, pace, days)
    }

/** Full net worth incl. super — a decades-scale forecast should count
 *  everything owned, whatever the Invest tab's toggle says today. */
val VestaStore.forecastNetWorth: Double
    get() = stocksValue + cryptoValue + debtNet

/** The effective levers: overrides win, measured values fill the gaps,
 *  the balanced preset is the last resort. */
val VestaStore.forecastInputs: ForecastInputs
    get() {
        val measured = measuredPace
        val a: ForecastAssumptions = forecastAssumptions
        val measuredDefault =
            if (measured.paceDays >= MEASURED_PACE_MIN_DAYS) measured.pacePct else null
        return ForecastInputs(
            netWorth = forecastNetWorth,
            monthlySaving = a.monthlySaving ?: measured.monthlySaving ?: 0.0,
            annualReturnPct = a.annualReturnPct ?: measuredDefault
                ?: ForecastMath.FALLBACK_RETURN_PCT,
            contributionGrowthPct = a.contributionGrowthPct,
        )
    }

/** The goal the forecast is aimed at: the nearest active one. */
val VestaStore.forecastGoal: NetworthGoal?
    get() = goals.filter { it.achievedAt == null }
        .minByOrNull { convert(it.amount, it.currency) }

/** Is the current return input exactly the measured pace? */
fun VestaStore.usingMeasuredReturn(inputs: ForecastInputs, measured: MeasuredPace): Boolean {
    val pace = measured.pacePct ?: return false
    return abs(inputs.annualReturnPct - pace) < 0.0001
}
